File: src/main/workspace-service.ts
import { opendir, open, realpath, stat } from 'node:fs/promises'
import { basename, extname, join } from 'node:path'
import type { ThreadRecord, WorkspaceFile, WorkspaceFilePreview } from '../shared/contracts'
import { isPathInside, resolveThreadPath } from './thread-path'

type ThreadLookup = (threadId: string) => ThreadRecord

interface ScannedFile extends WorkspaceFile {
  realPath: string
}

interface ScanCacheEntry {
  expiresAt: number
  files: ScannedFile[]
}

const MAX_FILES = 25_000
const MAX_DEPTH = 40
const MAX_PREVIEW_BYTES = 2 * 1024 * 1024
const CACHE_TTL_MS = 1_000
const IGNORED_DIRECTORIES = new Set([
  '.git',
  '.pi-gui',
  'node_modules',
  'out',
  'release',
  'dist',
  'coverage',
  '.next',
  '.turbo',
])

const LANGUAGE_BY_EXTENSION: Record<string, string> = {
  '.bash': 'shellscript',
  '.c': 'c',
  '.cc': 'cpp',
  '.cpp': 'cpp',
  '.css': 'css',
  '.fish': 'shellscript',
  '.go': 'go',
  '.h': 'c',
  '.hpp': 'cpp',
  '.htm': 'html',
  '.html': 'html',
  '.java': 'java',
  '.js': 'javascript',
  '.jsx': 'jsx',
  '.json': 'json',
  '.kt': 'kotlin',
  '.kts': 'kotlin',
  '.less': 'less',
  '.md': 'markdown',
  '.mdx': 'markdown',
  '.mjs': 'javascript',
  '.cjs': 'javascript',
  '.py': 'python',
  '.rs': 'rust',
  '.scss': 'scss',
  '.sh': 'shellscript',
  '.sql': 'sql',
  '.svelte': 'svelte',
  '.swift': 'swift',
  '.toml': 'toml',
  '.ts': 'typescript',
  '.tsx': 'tsx',
  '.vue': 'vue',
  '.xml': 'xml',
  '.yaml': 'yaml',
  '.yml': 'yaml',
  '.zsh': 'shellscript',
}

function languageForPath(path: string): string {
  const name = basename(path).toLowerCase()
  if (name === 'dockerfile') return 'dockerfile'
  if (name === 'makefile') return 'makefile'
  return LANGUAGE_BY_EXTENSION[extname(name)] ?? 'text'
}

function looksBinary(buffer: Buffer): boolean {
  const sampleLength = Math.min(buffer.length, 64 * 1024)
  if (sampleLength === 0) return false
  let suspicious = 0
  for (let index = 0; index < sampleLength; index += 1) {
    const byte = buffer[index]
    if (byte === 0) return true
    if (byte < 7 || (byte > 13 && byte < 32)) suspicious += 1
  }
  return suspicious / sampleLength > 0.08
}

function scoreFile(path: string, query: string): number {
  const normalizedPath = path.toLocaleLowerCase()
  const fileName = basename(normalizedPath)
  if (fileName === query) return 0
  if (fileName.startsWith(query)) return 1
  if (fileName.includes(query)) return 2
  if (normalizedPath.startsWith(query)) return 3
  return normalizedPath.includes(query) ? 4 : Number.POSITIVE_INFINITY
}

export class WorkspaceService {
  private readonly scanCache = new Map<string, ScanCacheEntry>()

  constructor(private readonly getThread: ThreadLookup) {}

  async listFiles(threadId: string): Promise<WorkspaceFile[]> {
    const files = await this.scan(threadId)
    return files.map(({ path, name, status }) => ({ path, name, ...(status ? { status } : {}) }))
  }

  async searchFiles(threadId: string, queryValue: string, limitValue = 40): Promise<WorkspaceFile[]> {
    const query = queryValue.trim().toLocaleLowerCase()
    if (!query) return []
    if (query.length > 1_000) throw new TypeError('File search query is too long')
    const limit = Math.min(100, Math.max(1, Math.trunc(limitValue)))
    const files = await this.scan(threadId)
    return files
      .map((file) => ({ file, score: scoreFile(file.path, query) }))
      .filter((item) => Number.isFinite(item.score))
      .sort((left, right) => left.score - right.score || left.file.path.length - right.file.path.length || left.file.path.localeCompare(right.file.path))
      .slice(0, limit)
      .map(({ file }) => ({ path: file.path, name: file.name }))
  }

  async recentFiles(threadId: string, limitValue = 12): Promise<WorkspaceFile[]> {
    const limit = Math.min(50, Math.max(1, Math.trunc(limitValue)))
    const files = await this.scan(threadId)
    const dated: Array<{ file: ScannedFile; modifiedAt: number }> = []
    const concurrency = 48
    for (let start = 0; start < files.length; start += concurrency) {
      const batch = files.slice(start, start + concurrency)
      const details = await Promise.all(batch.map(async (file) => ({
        file,
        modifiedAt: await stat(file.realPath).then((value) => value.mtimeMs).catch(() => 0),
      })))
      dated.push(...details)
    }
    return dated
      .filter((item) => item.modifiedAt > 0)
      .sort((left, right) => right.modifiedAt - left.modifiedAt)
      .slice(0, limit)
      .map(({ file }) => ({ path: file.path, name: file.name }))
  }

  async readFile(threadId: string, relativePath: string): Promise<WorkspaceFilePreview> {
    const thread = this.getThread(threadId)
    const resolved = await resolveThreadPath(thread, relativePath)
    const handle = await open(resolved.realPath, 'r')
    try {
      const details = await handle.stat()
      if (!details.isFile()) throw new Error('Workspace path is not a file')
      const bytesToRead = Math.min(details.size, MAX_PREVIEW_BYTES)
      const buffer = Buffer.allocUnsafe(bytesToRead)
      let offset = 0
      while (offset < bytesToRead) {
        const result = await handle.read(buffer, offset, bytesToRead - offset, offset)
        if (result.bytesRead === 0) break
        offset += result.bytesRead
      }
      const contentBuffer = offset === buffer.length ? buffer : buffer.subarray(0, offset)
      const binary = looksBinary(contentBuffer)
      const content = binary ? '' : contentBuffer.toString('utf8').replace(/^\uFEFF/, '')
      return {
        path: resolved.requestedPath,
        content,
        language: languageForPath(resolved.requestedPath),
        size: details.size,
        modifiedAt: details.mtimeMs,
        binary,
        truncated: details.size > contentBuffer.length,
      }
    } finally {
      await handle.close()
    }
  }

  invalidate(threadId?: string): void {
    if (threadId) this.scanCache.delete(threadId)
    else this.scanCache.clear()
  }

  private async scan(threadId: string): Promise<ScannedFile[]> {
    const cached = this.scanCache.get(threadId)
    if (cached && cached.expiresAt > Date.now()) return cached.files
    const thread = this.getThread(threadId)
    const root = await realpath(thread.cwd)
    const files: ScannedFile[] = []

    const walk = async (directory: string, prefix: string, depth: number): Promise<void> => {
      if (depth > MAX_DEPTH || files.length >= MAX_FILES) return
      const iterator = await opendir(directory)
      for await (const entry of iterator) {
        if (files.length >= MAX_FILES) break
        if (entry.name === '.DS_Store') continue
        const relativePath = prefix ? `${prefix}/${entry.name}` : entry.name
        const absolutePath = join(directory, entry.name)
        if (entry.isDirectory()) {
          if (!IGNORED_DIRECTORIES.has(entry.name)) await walk(absolutePath, relativePath, depth + 1)
          continue
        }
        if (entry.isFile()) {
          files.push({ path: relativePath, name: entry.name, realPath: absolutePath })
          continue
        }
        if (!entry.isSymbolicLink()) continue
        const target = await realpath(absolutePath).catch(() => undefined)
        if (!target || !isPathInside(root, target)) continue
        const targetDetails = await stat(target).catch(() => undefined)
        // Do not recurse through directory symlinks: it avoids cycles and keeps
        // file enumeration rooted in the visible workspace tree.
        if (targetDetails?.isFile()) files.push({ path: relativePath, name: entry.name, realPath: target })
      }
    }

    await walk(root, '', 0)
    files.sort((left, right) => left.path.localeCompare(right.path, 'en', { numeric: true, sensitivity: 'base' }))
    this.scanCache.set(threadId, { expiresAt: Date.now() + CACHE_TTL_MS, files })
    return files
  }
}

File: macos/Sources/CodePiKit/WorkspaceService.swift
import Foundation

/// Port of `thread-path.ts` + `workspace-service.ts`: safe workspace file
/// enumeration and bounded read-only previews.
public enum ThreadPath {
  public static func normalizeRelativePath(_ value: String, allowRoot: Bool = false) throws -> String {
    guard !value.contains("\0"), value.count <= 16_384 else { throw BridgeError("Workspace path is invalid") }
    let portable = value.replacingOccurrences(of: "\\", with: "/")
    if portable.isEmpty {
      if allowRoot { return "" }
      throw BridgeError("Workspace path cannot be empty")
    }
    if portable.hasPrefix("/") || portable.range(of: "^[A-Za-z]:", options: .regularExpression) != nil {
      throw BridgeError("Workspace path must be relative")
    }
    let pieces = portable.components(separatedBy: "/")
    if pieces.contains(where: { $0.isEmpty || $0 == "." || $0 == ".." }) {
      throw BridgeError("Workspace path contains an unsafe segment")
    }
    return pieces.joined(separator: "/")
  }

  public static func isInside(root: String, target: String) -> Bool {
    let rootPath = URL(fileURLWithPath: root).standardizedFileURL.path
    let targetPath = URL(fileURLWithPath: target).standardizedFileURL.path
    return targetPath == rootPath || targetPath.hasPrefix(rootPath + "/")
  }

  /// Resolve an existing path and verify both its lexical and real path stay
  /// in the thread root.
  public static func resolve(cwd: String, requestedPath: String, allowRoot: Bool = false) throws -> (root: String, requestedPath: String, realPath: String) {
    let normalized = try normalizeRelativePath(requestedPath, allowRoot: allowRoot)
    let root = URL(fileURLWithPath: cwd).resolvingSymlinksInPath().standardizedFileURL.path
    let lexical = normalized.isEmpty
      ? root
      : URL(fileURLWithPath: root).appendingPathComponent(normalized).standardizedFileURL.path
    guard isInside(root: root, target: lexical) else { throw BridgeError("Workspace path escapes the thread directory") }
    let real = URL(fileURLWithPath: lexical).resolvingSymlinksInPath().standardizedFileURL.path
    guard isInside(root: root, target: real) else { throw BridgeError("Workspace symlink points outside the thread directory") }
    return (root, normalized, real)
  }
}

@MainActor
public final class WorkspaceService {
  public struct ScannedFile: Sendable {
    public let path: String
    public let name: String
    public let realPath: String
  }

  private struct CacheEntry {
    let expiresAt: Date
    let files: [ScannedFile]
  }

  static let maxFiles = 25_000
  static let maxDepth = 40
  static let maxPreviewBytes = 2 * 1024 * 1024
  static let ignoredDirectories: Set<String> = [
    ".git", ".pi-gui", "node_modules", "out", "release", "dist", "coverage", ".next", ".turbo"
  ]

  static let languageByExtension: [String: String] = [
    "bash": "shellscript", "c": "c", "cc": "cpp", "cpp": "cpp", "css": "css",
    "fish": "shellscript", "go": "go", "h": "c", "hpp": "cpp", "htm": "html",
    "html": "html", "java": "java", "js": "javascript", "jsx": "jsx", "json": "json",
    "kt": "kotlin", "kts": "kotlin", "less": "less", "md": "markdown", "mdx": "markdown",
    "mjs": "javascript", "cjs": "javascript", "py": "python", "rs": "rust", "scss": "scss",
    "sh": "shellscript", "sql": "sql", "svelte": "svelte", "swift": "swift", "toml": "toml",
    "ts": "typescript", "tsx": "tsx", "vue": "vue", "xml": "xml", "yaml": "yaml",
    "yml": "yaml", "zsh": "shellscript"
  ]

  private var scanCache: [String: CacheEntry] = [:]

  public init() {}

  public static func languageForPath(_ path: String) -> String {
    let name = (path as NSString).lastPathComponent.lowercased()
    if name == "dockerfile" { return "dockerfile" }
    if name == "makefile" { return "makefile" }
    return languageByExtension[(name as NSString).pathExtension] ?? "text"
  }

  public static func looksBinary(_ data: Data) -> Bool {
    let sample = data.prefix(64 * 1024)
    if sample.isEmpty { return false }
    var suspicious = 0
    for byte in sample {
      if byte == 0 { return true }
      if byte < 7 || (byte > 13 && byte < 32) { suspicious += 1 }
    }
    return Double(suspicious) / Double(sample.count) > 0.08
  }

  public func invalidate(threadId: String? = nil) {
    if let threadId { scanCache.removeValue(forKey: threadId) } else { scanCache.removeAll() }
  }

  public func listFiles(threadId: String, cwd: String) throws -> JSONValue {
    let files = try scan(threadId: threadId, cwd: cwd)
    return .array(files.map { .object(["path": .string($0.path), "name": .string($0.name)]) })
  }

  public func searchFiles(threadId: String, cwd: String, query rawQuery: String, limit rawLimit: Int) throws -> JSONValue {
    let query = rawQuery.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    if query.isEmpty { return .array([]) }
    guard query.count <= 1_000 else { throw BridgeError("File search query is too long") }
    let limit = min(100, max(1, rawLimit))
    let files = try scan(threadId: threadId, cwd: cwd)
    let scored = files.compactMap { file -> (file: ScannedFile, score: Int)? in
      let path = file.path.lowercased()
      let name = file.name.lowercased()
      let score: Int
      if name == query { score = 0 }
      else if name.hasPrefix(query) { score = 1 }
      else if name.contains(query) { score = 2 }
      else if path.hasPrefix(query) { score = 3 }
      else if path.contains(query) { score = 4 }
      else { return nil }
      return (file, score)
    }
    let sorted = scored.sorted {
      $0.score != $1.score
        ? $0.score < $1.score
        : ($0.file.path.count != $1.file.path.count ? $0.file.path.count < $1.file.path.count : $0.file.path < $1.file.path)
    }
    return .array(sorted.prefix(limit).map { .object(["path": .string($0.file.path), "name": .string($0.file.name)]) })
  }

  public func recentFiles(threadId: String, cwd: String, limit rawLimit: Int = 12) throws -> JSONValue {
    let limit = min(50, max(1, rawLimit))
    let files = try scan(threadId: threadId, cwd: cwd)
    let manager = FileManager.default
    let dated = files.compactMap { file -> (file: ScannedFile, modified: Date)? in
      guard let date = (try? manager.attributesOfItem(atPath: file.realPath))?[.modificationDate] as? Date else {
        return nil
      }
      return (file, date)
    }
    return .array(dated.sorted { $0.modified > $1.modified }.prefix(limit).map {
      .object(["path": .string($0.file.path), "name": .string($0.file.name)])
    })
  }

  public func readFile(cwd: String, relativePath: String) throws -> JSONValue {
    let resolved = try ThreadPath.resolve(cwd: cwd, requestedPath: relativePath)
    guard let handle = FileHandle(forReadingAtPath: resolved.realPath) else {
      throw BridgeError("Workspace file could not be opened")
    }
    defer { try? handle.close() }
    let attributes = try FileManager.default.attributesOfItem(atPath: resolved.realPath)
    guard (attributes[.type] as? FileAttributeType) == .typeRegular else {
      throw BridgeError("Workspace path is not a file")
    }
    let size = (attributes[.size] as? Int) ?? 0
    let modified = ((attributes[.modificationDate] as? Date)?.timeIntervalSince1970 ?? 0) * 1000
    let data = (try? handle.read(upToCount: Self.maxPreviewBytes)) ?? Data()
    let binary = Self.looksBinary(data)
    var content = binary ? "" : String(decoding: data, as: UTF8.self)
    if content.hasPrefix("\u{FEFF}") { content.removeFirst() }
    return .object([
      "path": .string(resolved.requestedPath),
      "content": .string(content),
      "language": .string(Self.languageForPath(resolved.requestedPath)),
      "size": .number(Double(size)),
      "modifiedAt": .number(modified),
      "binary": .bool(binary),
      "truncated": .bool(size > data.count)
    ])
  }

  private func scan(threadId: String, cwd: String) throws -> [ScannedFile] {
    if let cached = scanCache[threadId], cached.expiresAt > Date() { return cached.files }
    let root = URL(fileURLWithPath: cwd).resolvingSymlinksInPath().standardizedFileURL.path
    var files: [ScannedFile] = []
    let manager = FileManager.default

    func walk(directory: String, prefix: String, depth: Int) {
      if depth > Self.maxDepth || files.count >= Self.maxFiles { return }
      let entries = (try? manager.contentsOfDirectory(atPath: directory)) ?? []
      for name in entries {
        if files.count >= Self.maxFiles { break }
        if name == ".DS_Store" { continue }
        let relativePath = prefix.isEmpty ? name : "\(prefix)/\(name)"
        let absolute = directory + "/" + name
        var isSymlink = false
        if let type = (try? manager.attributesOfItem(atPath: absolute))?[.type] as? FileAttributeType {
          isSymlink = type == .typeSymbolicLink
          if type == .typeDirectory {
            if !Self.ignoredDirectories.contains(name) {
              walk(directory: absolute, prefix: relativePath, depth: depth + 1)
            }
            continue
          }
          if type == .typeRegular {
            files.append(ScannedFile(path: relativePath, name: name, realPath: absolute))
            continue
          }
        }
        guard isSymlink else { continue }
        let target = URL(fileURLWithPath: absolute).resolvingSymlinksInPath().standardizedFileURL.path
        guard ThreadPath.isInside(root: root, target: target) else { continue }
        var isDirectory: ObjCBool = false
        // Directory symlinks are not traversed: avoids cycles, keeps
        // enumeration rooted in the visible workspace tree.
        if manager.fileExists(atPath: target, isDirectory: &isDirectory), !isDirectory.boolValue {
          files.append(ScannedFile(path: relativePath, name: name, realPath: target))
        }
      }
    }

    walk(directory: root, prefix: "", depth: 0)
    files.sort { $0.path.localizedStandardCompare($1.path) == .orderedAscending }
    scanCache[threadId] = CacheEntry(expiresAt: Date().addingTimeInterval(1), files: files)
    return files
  }
}
